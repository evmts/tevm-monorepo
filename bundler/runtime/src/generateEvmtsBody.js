import { generateDtsBody } from './generateEvmtsBodyDts.js'
import { formatAbi } from 'abitype'
import { succeed } from 'effect/Effect'

/**
 * @param {import("@evmts/solc").Artifacts} artifacts
 * @param {import('./types.js').ModuleType} moduleType
 * @param {boolean} includeBytecode
 * @returns {import('effect/Effect').Effect<never, never, string>}
 */
export const generateEvmtsBody = (artifacts, moduleType, includeBytecode) => {
	if (moduleType === 'dts') {
		return generateDtsBody(artifacts, includeBytecode)
	}
	return succeed(
		Object.entries(artifacts)
			.flatMap(([contractName, { abi, userdoc = {}, evm }]) => {
				const contract = JSON.stringify({
					name: contractName,
					humanReadableAbi: formatAbi(abi),
					bytecode: evm?.bytecode.object && `0x${evm.bytecode.object}`,
					deployedBytecode:
						evm?.deployedBytecode.object && `0x${evm.deployedBytecode.object}`,
				})
				const natspec = Object.entries(userdoc.methods ?? {}).map(
					([method, { notice }]) => ` * @property ${method} ${notice}`,
				)
				if (userdoc.notice) {
					natspec.unshift(` * ${userdoc.notice}`)
				}
				if (natspec.length) {
					natspec.unshift('/**')
					natspec.push(' */')
				}
				if (moduleType === 'cjs') {
					return [
						`const _${contractName} = ${contract}`,
						...natspec,
						`module.exports.${contractName} = evmtsContractFactory(_${contractName})`,
					]
				}

				if (moduleType === 'ts') {
					return [
						`const _${contractName} = ${contract} as const`,
						...natspec,
						`export const ${contractName} = evmtsContractFactory(_${contractName})`,
					]
				}

				return [
					`const _${contractName} = ${contract}`,
					...natspec,
					`export const ${contractName} = evmtsContractFactory(_${contractName})`,
				]
			})
			.join('\n'),
	)
}
