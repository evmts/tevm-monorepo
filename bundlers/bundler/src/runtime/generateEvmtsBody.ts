import type { Artifacts } from '../solc/resolveArtifactsSync'
import { generateDtsBody } from './generateEvmtsBodyDts'
import type { ResolvedConfig } from '@evmts/config'

type ModuleType = 'cjs' | 'mjs' | 'ts' | 'dts'

export const generateEvmtsBody = (
	artifacts: Artifacts,
	config: ResolvedConfig,
	moduleType: ModuleType,
): string => {
	if (moduleType === 'dts') {
		return generateDtsBody(artifacts, config)
	}
	return Object.entries(artifacts)
		.flatMap(([contractName, { abi, userdoc = {} }]) => {
			const contract = JSON.stringify({
				name: contractName,
				abi,
				addresses:
					config.localContracts.contracts?.find(
						(contractConfig) => contractConfig.name === contractName,
					)?.addresses ?? {},
			})

			const natspec = Object.entries(userdoc.methods ?? {}).map(
				([method, { notice }]) => ` * @property ${method} ${notice}`,
			)
			if (userdoc.notice) {
				natspec.unshift(` * ${userdoc.notice}`)
			}
			if (natspec.length) {
				natspec.unshift('/**')
				natspec.push(' */')
			}
			if (moduleType === 'cjs') {
				return [
					`const _${contractName} = ${contract}`,
					...natspec,
					`module.exports.${contractName} = evmtsContractFactory(_${contractName})`,
				]
			}

			if (moduleType === 'ts') {
				return [
					`const _${contractName} = ${contract} as const`,
					...natspec,
					`export const ${contractName} = evmtsContractFactory(_${contractName})`,
				]
			}

			return [
				`const _${contractName} = ${contract}`,
				...natspec,
				`export const ${contractName} = evmtsContractFactory(_${contractName})`,
			]
		})
		.join('\n')
}
