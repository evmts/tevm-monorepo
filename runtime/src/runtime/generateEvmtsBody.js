import { generateDtsBody } from './generateEvmtsBodyDts.js'
import { formatAbi } from 'abitype'

/**
 * @typedef {'cjs' | 'mjs' | 'ts' | 'dts'} ModuleType
 */

/**
 * @param {import("../solc/resolveArtifactsSync.js").Artifacts} artifacts
 * @param {ModuleType} moduleType
 * @returns {string}
 */
export const generateEvmtsBody = (artifacts, moduleType) => {
	if (moduleType === 'dts') {
		return generateDtsBody(artifacts)
	}
	return Object.entries(artifacts)
		.flatMap(([contractName, { abi, userdoc = {} }]) => {
			const contract = JSON.stringify({
				name: contractName,
				humanReadableAbi: formatAbi(abi),
			})

			const natspec = Object.entries(userdoc.methods ?? {}).map(
				([method, { notice }]) => ` * @property ${method} ${notice}`,
			)
			if (userdoc.notice) {
				natspec.unshift(` * ${userdoc.notice}`)
			}
			if (natspec.length) {
				natspec.unshift('/**')
				natspec.push(' */')
			}
			if (moduleType === 'cjs') {
				return [
					`const _${contractName} = ${contract}`,
					...natspec,
					`module.exports.${contractName} = evmtsContractFactory(_${contractName})`,
				]
			}

			if (moduleType === 'ts') {
				return [
					`const _${contractName} = ${contract} as const`,
					...natspec,
					`export const ${contractName} = evmtsContractFactory(_${contractName})`,
				]
			}

			return [
				`const _${contractName} = ${contract}`,
				...natspec,
				`export const ${contractName} = evmtsContractFactory(_${contractName})`,
			]
		})
		.join('\n')
}
