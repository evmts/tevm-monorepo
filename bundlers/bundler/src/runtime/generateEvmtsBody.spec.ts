import { generateEvmtsBody } from './generateEvmtsBody' // replace with your actual module path
import { describe, expect, it } from 'vitest'

describe('generateEvmtsBody', () => {
	const artifacts = {
		MyContract: {
			abi: {},
			bytecode: '0x60016001',
		},
		AnotherContract: {
			abi: {},
			bytecode: '0x60016002',
		},
	}

	const config = {
		localContracts: {
			contracts: [
				{
					name: 'MyContract',
					addresses: {
						test: '0x123',
					},
				},
			],
		},
	}

	it('should generate correct body for cjs module', () => {
		const result = generateEvmtsBody(
			artifacts,
			config as any,
			'cjs',
		)
		expect(result).toMatchSnapshot()
	})

	it('should generate correct body for mjs module', () => {
		const result = generateEvmtsBody(
			artifacts,
			config as any,
			'mjs',
		)
		expect(result).toMatchSnapshot()
	})

	it('should generate correct body for ts module', () => {
		const result = generateEvmtsBody(
			artifacts,
			config as any,
			'ts',
		)
		expect(result).toMatchSnapshot()
	})
})
