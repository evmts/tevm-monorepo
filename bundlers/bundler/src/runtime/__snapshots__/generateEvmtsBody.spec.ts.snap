// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateEvmtsBody > should generate correct body for cjs module 1`] = `
"const _MyContract = {\\"name\\":\\"MyContract\\",\\"abi\\":{},\\"bytecode\\":\\"0x60016001\\",\\"addresses\\":{\\"test\\":\\"0x123\\"}}
module.exports.MyContract = evmtsContractFactory(_MyContract)
const _AnotherContract = {\\"name\\":\\"AnotherContract\\",\\"abi\\":{},\\"bytecode\\":\\"0x60016002\\",\\"addresses\\":{}}
module.exports.AnotherContract = evmtsContractFactory(_AnotherContract)"
`;

exports[`generateEvmtsBody > should generate correct body for mjs module 1`] = `
"const _MyContract = {\\"name\\":\\"MyContract\\",\\"abi\\":{},\\"bytecode\\":\\"0x60016001\\",\\"addresses\\":{\\"test\\":\\"0x123\\"}}
export const MyContract = evmtsContractFactory(_MyContract)
const _AnotherContract = {\\"name\\":\\"AnotherContract\\",\\"abi\\":{},\\"bytecode\\":\\"0x60016002\\",\\"addresses\\":{}}
export const AnotherContract = evmtsContractFactory(_AnotherContract)"
`;

exports[`generateEvmtsBody > should generate correct body for ts module 1`] = `
"const _MyContract = {\\"name\\":\\"MyContract\\",\\"abi\\":{},\\"bytecode\\":\\"0x60016001\\",\\"addresses\\":{\\"test\\":\\"0x123\\"}} as const
export const MyContract = evmtsContractFactory(_MyContract)
const _AnotherContract = {\\"name\\":\\"AnotherContract\\",\\"abi\\":{},\\"bytecode\\":\\"0x60016002\\",\\"addresses\\":{}} as const
export const AnotherContract = evmtsContractFactory(_AnotherContract)"
`;
